crates/scsi/src/scsi/emulation/block_device.rs

```diff
@@ -0,0 +1,547 @@
+                                                   // don't support anchored
+                                                   // LBAs or group descriptors
+                            out.push(0b0000_0010); // thin provisioned
+                            out.push(0); // no threshold % support
+                        }
+                        _ => return Ok(CmdOutput::check_condition(sense::INVALID_FIELD_IN_CDB)),
+                    }
+                    data_in.write_all(&[code.into()]).map_err(DataIn)?;
+                    // unwrap is fine: none of our VPD pages are over 2^16 bits long
+                    data_in
+                        .write_all(&u16::try_from(out.len()).unwrap().to_be_bytes())
+                        .map_err(DataIn)?;
+                    data_in.write_all(&out).map_err(DataIn)?;
+                } else {
+                    respond_standard_inquiry_data(&mut data_in).map_err(DataIn)?;
+                }
+
+                Ok(CmdOutput::ok())
+            }
+            Command::ReportSupportedOperationCodes { rctd, mode } => {
+                // helpers for output data format
+                fn one_command_supported(
+                    data_in: &mut impl Write,
+                    ty: CommandType,
+                ) -> io::Result<()> {
+                    data_in.write_all(&[0])?; // unused flags
+                    data_in.write_all(&[0b0000_0011])?; // supported, don't set a bunch of flags
+                    let tpl = ty.cdb_template();
+                    // unwrap is safe: the length of the template is the length of the CDB, and no
+                    // command we support is anywhere near that long
+                    data_in.write_all(&u16::try_from(tpl.len()).unwrap().to_be_bytes())?;
+                    data_in.write_all(tpl)?;
+                    Ok(())
+                }
+                fn one_command_not_supported(data_in: &mut impl Write) -> io::Result<()> {
+                    data_in.write_all(&[0])?; // unused flags
+                    data_in.write_all(&[0b0000_0001])?; // not supported
+                    data_in.write_all(&[0; 2])?; // cdb len
+                    Ok(())
+                }
+                fn timeout_descriptor(data_in: &mut impl Write) -> io::Result<()> {
+                    // timeout descriptor
+                    data_in.write_all(&0xa_u16.to_be_bytes())?; // len
+                    data_in.write_all(&[0, 0])?; // reserved, cmd specific
+                    data_in.write_all(&0_u32.to_be_bytes())?;
+                    data_in.write_all(&0_u32.to_be_bytes())?;
+                    Ok(())
+                }
+
+                match mode {
+                    ReportSupportedOpCodesMode::All => {
+                        let cmd_len = if rctd { 20 } else { 8 };
+                        // unwrap is safe: we're never going to have (2^32 / 20) ~= 2^27 opcodes
+                        let len = u32::try_from(OPCODES.len() * cmd_len).unwrap();
```

Comment:
in general, many of the unwrap() comments could probably turned into a .expect()